### DISCUSSION TITLE
Dynamic steps

### DISCUSSION THREAD

**Discussion by pkernevez** [spring-projects/spring-batch#4662] [Q&A] [answered]
Hi,

We have a Job that is supposed to be run every day. It has multiple steps.
One of the step (named StepA) reuse the result of the previous day as an input for is own execution.
As for now stepA have one parameter: the day we are processing.
This step is quite complexe, and use chuck parallel processing. It's working well.

Now we want to be able to deal with a production incident that miss an ore several executions.
I would like to be able to define dynamically the steps, to decide before running the job that for this job we need to run:
* stepA for 2 days ago
* stepA for 1 day ago
* stepA today
* other steps...

How could we define this list dynamically ?

* I tried to use Flows, but I don't see how keep a context for the counter to do the loop.

* What do you think about defining the Job with a prototype scope and changing the steps contextually depending on the missing days ?

Thanks for your help,

**Comment by fmbenhassine** [comment, marked as answer]
Hi,

> I would like to be able to define dynamically the steps, to decide before running the job that for this job we need to run

You can implement the logic of that decision in a decider. Here is a quick example: https://gist.github.com/fmbenhassine/3dd9ecf69bb1dafe6664aed5ab021099. The example starts with an initial step, but we added the ability to start a job with a decider in v5.1 (see #4411). Once the decision is made, then you can define the steps dynamically with something like this: [How to create dynamic steps in Spring Batch](https://stackoverflow.com/questions/54853908/how-to-create-dynamic-steps-in-spring-batch).

Let me know if this helps.

**Reply by pkernevez** [reply]
I finally did it with a Decider and a Flow.
The decider to decide if we want to go in the Loop and the Flow to do the iterations and manage the JobContex.

**Reply by fmbenhassine** [reply]
Perfect, thank you for letting me know!
